[AutoDiff] Differentiation transform cleanup.

dan-zheng opened the pull request:
- Remove all mentions of "primal".
  - Rename "primal value struct" to "pullback struct".
- Simplify/optimize `VJPEmitter` and `AdjointEmitter` members.
  - Unuseful `VJPEmitter` getter methods are removed.
  - `AdjointEmitter` now stores a reference to the parent `VJPEmitter`.
    `AdjointEmitter` getter methods forward `VJPEmitter` members.
- Use `SmallPtrSet` to track processed `autodiff_function` instructions
  to avoid reprocessing invalidated instructions.
- Redefine `foldAutoDiffFunctionExtraction` as a method on `ADContext`.

dan-zheng committed:
[AutoDiff] Differentiation transform cleanup.

- Remove all mentions of "primal".
  - Rename "primal value struct" to "pullback struct".
- Simplify/optimize `VJPEmitter` and `AdjointEmitter` members.
  - Unuseful `VJPEmitter` getter methods are removed.
  - `AdjointEmitter` now stores a reference to the parent `VJPEmitter`.
    `AdjointEmitter` getter methods forward `VJPEmitter` members.
- Use `SmallPtrSet` to track processed `autodiff_function` instructions
  to avoid reprocessing invalidated instructions.
- Redefine `foldAutoDiffFunctionExtraction` as a method on `ADContext`.

dan-zheng commented:
More cleanup coming soon. I'm splitting it out of https://github.com/apple/swift/pull/24911.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -666,7 +666,7 @@ class ADContext {
   Lowering::TypeConverter &getTypeConverter() { return module.Types; }
 
-  SmallVectorImpl<AutoDiffFunctionInst *> &getAutoDiffFunctionInsts() {
+  SmallSetVector<AutoDiffFunctionInst *, 32> &getAutoDiffFunctionInsts() {
```
```suggestion
  SetVector<AutoDiffFunctionInst *> &getAutoDiffFunctionInsts() {
```

dan-zheng replied:
Using `SmallVector` + `SmallPtrSet` instead.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -5982,11 +5990,8 @@ static void foldAutoDiffFunctionExtraction(
-    if (inst == source)
-      inst = nullptr;
+  // Remove `source` from the worklist.
+  getAutoDiffFunctionInsts().remove(source);
```
Removal from `SetVector<T>` is `O(n)`: http://llvm.org/docs/ProgrammersManual.html#llvm-adt-setvector-h. If you can, still, rewrite the element to `nullptr`, that would be good.  Otherwise, the best solution would be to revert it to `SmallVector` and use a separate `SmallPtrSet` to track pointers that have already been removed.

dan-zheng replied:
Mutating `SetVector` elements doesn't seem to work:
```diff
@@ -5990,8 +5958,11 @@ void ADContext::foldAutoDiffFunctionExtraction(AutoDiffFunctionInst *source) {
   if (!isInstructionTriviallyDead(source))
     return;
   source->eraseFromParent();
-  // Remove `source` from the worklist.
-  getAutoDiffFunctionInsts().remove(source);
+  // Delete all worklist occurrences of `adfi` by setting them to nullptr.
+  // This is more efficient than APIs like `llvm::erase_if`.
+  for (auto &inst : getAutoDiffFunctionInsts())
+    if (inst == source)
+      inst = nullptr;
 }
```

```
/Users/danielzheng/swift-merge/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:5965:12: error: cannot assign to variable 'inst' with const-qualified type 'swift::AutoDiffFunctionInst *const &'
      inst = nullptr;
      ~~~~ ^
/Users/danielzheng/swift-merge/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:5963:14: note: variable 'inst' declared const here
  for (auto &inst : getAutoDiffFunctionInsts())
       ~~~~~~^~~~~~
1 error generated.
```

I'll follow your suggestion of `SmallVector` + `SmallPtrSet`!

dan-zheng replied:
Done in https://github.com/apple/swift/pull/24946/commits/eba1c455d12732d7892430cff19c6dc0b8f691e3.

```diff
+  /// The set of `autodiff_function` instructions that have been processed.
+  /// Used to avoid reprocessing invalidated instructions.
+  SmallPtrSet<AutoDiffFunctionInst *, 32> processedAutoDiffFunctionInsts;

@@ -5982,11 +5955,8 @@ static void foldAutoDiffFunctionExtraction(
   if (!isInstructionTriviallyDead(source))
     return;
   source->eraseFromParent();
-  // Delete all worklist occurrences of `source` by setting them to nullptr.
-  // This is more efficient than APIs like `llvm::erase_if`.
-  for (auto &inst : context.getAutoDiffFunctionInsts())
-    if (inst == source)
-      inst = nullptr;
+  // Mark `source` as processed so that it won't be reprocessed after deletion.
+  processedAutoDiffFunctionInsts.insert(source);

@@ -6080,9 +6047,8 @@ void Differentiation::run() {
   while (!context.getAutoDiffFunctionInsts().empty()) {
     auto *adfi = context.getAutoDiffFunctionInsts().back();
     context.getAutoDiffFunctionInsts().pop_back();
-    // Skip instructions that have been set to nullptr by
-    // `processAutoDiffFunctionInst`.
-    if (!adfi) continue;
+    // Skip instructions that have been already been processed.
+    if (context.getProcessedAutoDiffFunctionInsts().count(adfi)) continue;
     errorOccurred |= context.processAutoDiffFunctionInst(adfi);
   }
```

dan-zheng commented:
Ready for re-review. cc @rxwei

The new changes are:
* The `autodiff_function` worklist is reverted to a `SmallVector` and a `SmallPtrSet` keeps track of previously processed instructions.
* All mentions of "primal" have been removed. "Primal value struct" is now referred to as "pullback struct".
* `VJPEmitter` and `AdjointEmitter` members are simplified. `AdjointEmitter` stores a reference to its parent `VJPEmitter` and forwards its members.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2590,25 +2606,14 @@ class VJPEmitter final
+  /// List of pullbacks, used to build the pullback struct.
+  SmallVector<SILValue, 8> pullbacks;
 
+  ADContext &getContext() const { return context; }
```
I don't think we need this one either.

dan-zheng replied:
Good catch, thanks!

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
I'm not sure why so many tests are failing:
```
Failing Tests (22):
    Swift(linux-x86_64) :: AutoDiff/existential.swift
    Swift(linux-x86_64) :: AutoDiff/autodiff_diagnostics.swift
    Swift(linux-x86_64) :: AutoDiff/refcounting.swift
    Swift(linux-x86_64) :: TensorFlowRuntime/tensor_autodiff_indirect.swift
    Swift(linux-x86_64) :: TensorFlowRuntime/model_autodiff_runtime.swift
    Swift(linux-x86_64) :: TensorFlowRuntime/tensor_autodiff_runtime.swift
    Swift(linux-x86_64) :: AutoDiff/protocol_requirement_autodiff.swift
    Swift(linux-x86_64) :: AutoDiff/simple_math.swift
    Swift(linux-x86_64) :: AutoDiff/superset_adjoint.swift
    Swift(linux-x86_64) :: AutoDiff/leakchecking.swift
    Swift(linux-x86_64) :: AutoDiff/generics.swift
    Swift(linux-x86_64) :: AutoDiff/simple_model.swift
    Swift(linux-x86_64) :: AutoDiff/repeated_calls.swift
    Swift(linux-x86_64) :: AutoDiff/e2e_differentiable_property.swift
    Swift(linux-x86_64) :: AutoDiff/method.swift
    Swift(linux-x86_64) :: AutoDiff/array.swift
    Swift(linux-x86_64) :: AutoDiff/anyderivative.swift
    Swift(linux-x86_64) :: AutoDiff/custom_derivatives.swift
    Swift(linux-x86_64) :: AutoDiff/builtin_differential_operators.swift
    Swift(linux-x86_64) :: AutoDiff/differential_operators.swift
    Swift(linux-x86_64) :: AutoDiff/closures.swift
    Swift(linux-x86_64) :: AutoDiff/derivative_registration.swift
```

All seem due to this verification error (empty `autodiff_function` instruction):
```
swift: /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/swift/lib/SIL/SILInstructions.cpp:611: std::pair<SILValue, SILValue> swift::AutoDiffFunctionInst::getAssociatedFunctionPair(unsigned int) const: Assertion `!getAssociatedFunctions().empty() && "No associated functions. Maybe " "the differentiation pass has not run?"' failed.
```
I can't reproduce the test failures on macOS: stdlib compiles fine and tests pass.

dan-zheng committed:
Fix worklist logic, gardening.

- Add `ADContext::createAutoDiffFunction`, which constructs a new
  `autodiff_function` instruction and removes it from the processed
  instruction set if it exists (it may share the pointer value as a
  previously processed and deleted instruction).

dan-zheng commented:
@rxwei: `autodiff_function` worklist fix in https://github.com/apple/swift/pull/24946/commits/dde15dcac0d456cc4c65da519eea3841e38ffb70 is ready for review.

Newly created `autodiff_function` instructions can share the same pointer value as previously processed and deleted ones, so it's necessary to erase newly created instructions from the processed instruction set.

This approach should still be more efficient than the old `O(n)` approach, i.e. invalidation by setting to nullptr.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2566,21 +2592,26 @@ static SILFunction *getAssociatedFunction(
+  SILFunction *adjoint;
+
+  /// The pullback info.
+  std::unique_ptr<PullbackInfo> pullbackInfo = nullptr;
```
hmm why is this still a unique ptr?

dan-zheng replied:
Nice catch 🙂 
That will be fixed in https://github.com/apple/swift/pull/24911. Changing that in this PR was difficult for some reason.

dan-zheng replied:
My mistake - there's no reason why `PullbackInfo` can't be stored as a normal value instead of `unique_ptr`. I forgot to apply that change in this PR, it'll be fixed in my next patch.
